src/table/Source.cc

```diff
@@ -407,38 +316,127 @@ PTR(BaseRecord) SourceFitsReader::_readRecord(PTR(BaseTable) const & table) {
                 fp->addPeak(x, y, value);
             }
         }
-        record->setFootprint(fp);
+        record.setFootprint(fp);
 
-        if (
-            !(_flags & SOURCE_IO_NO_HEAVY_FOOTPRINTS)
-            && heavyPixElementCount && heavyMaskElementCount && heavyVarElementCount
-        ) {
+        // If we're setup to read HeavyFootprints
+        if (_heavyPixCol < 0) {
+            return;
+        }
+        int heavyPixElementCount  = fits.getTableArraySize(row, _heavyPixCol);
+        int heavyMaskElementCount = fits.getTableArraySize(row, _heavyMaskCol);
+        int heavyVarElementCount  = fits.getTableArraySize(row, _heavyVarCol);
+        if (heavyPixElementCount > 0) {
             int N = fp->getArea();
-            if ((heavyPixElementCount  != N) ||
-                (heavyMaskElementCount != N) ||
-                (heavyVarElementCount  != N)) {
+            if ((heavyPixElementCount != N) || (heavyMaskElementCount != N) || (heavyVarElementCount != N)) {
                 throw LSST_EXCEPT(
                     afw::fits::FitsError,
                     afw::fits::makeErrorMessage(
-                        _fits->fptr, _fits->status,
-                        boost::format("Number of HeavyFootprint elements (pix %d, mask %d, var %d) must all be equal to footprint area (%d)")
+                        fits.fptr, fits.status,
+                        boost::format("Number of HeavyFootprint elements (pix %d, mask %d, var %d) "
+                                      "must all be equal to footprint area (%d)")
                         % heavyPixElementCount % heavyMaskElementCount % heavyVarElementCount % N
-                        ));
+                    )
+                );
             }
             // float HeavyFootprints were the only kind we ever saved using the old format
             typedef detection::HeavyFootprint<float,image::MaskPixel,image::VariancePixel> HeavyFootprint;
             PTR(HeavyFootprint) heavy = boost::make_shared<HeavyFootprint>(*fp);
-            _fits->readTableArray(_row, _heavyPixCol,  N, heavy->getImageArray().getData());
-            _fits->readTableArray(_row, _heavyMaskCol, N, heavy->getMaskArray().getData());
-            _fits->readTableArray(_row, _heavyVarCol,  N, heavy->getVarianceArray().getData());
-            record->setFootprint(heavy);
+            fits.readTableArray(row, _heavyPixCol,  N, heavy->getImageArray().getData());
+            fits.readTableArray(row, _heavyMaskCol, N, heavy->getMaskArray().getData());
+            fits.readTableArray(row, _heavyVarCol,  N, heavy->getVarianceArray().getData());
+            record.setFootprint(heavy);
         }
     }
-    return record;
-}
+
+private:
+    int _spanCol;
+    int _peakCol;
+    int _heavyPixCol;
+    int _heavyMaskCol;
+    int _heavyVarCol;
+};
+
+// FitsColumnReader for new-style Footprint persistence using archives.
+class SourceFootprintReader : public io::FitsColumnReader {
```

Comment:
Yeah, I thought about trying to share some code here, but there simply wasn't any real code to share - it's just boilerplate that has to be repeated anyway.